docs/decisions/0002-lerna-monorepo.rst

```diff
@@ -0,0 +1,108 @@
+
+When a branch is pushed or a pull request is created on Github, CI will continue to run tests, linting, and other necessary build steps to ensure changes did not cause regressions. The difference now is that we rely on Lerna to run commands across all packages that have changed since the last release. This will ensure when a change is made to a package, the CI will be sure to run any associated tests for that specific package, including any of its dependencies and dependents across the monorepo. Output from each command Lerna runs concurrently across all packages are streamed in each Github Action step.
+
+================================
+NPM releases with Lerna
+================================
+
+Publishing packages to NPM remains a job of Github Actions, however, releases will no longer happen using ``semantic-release`` directly. Instead, we will rely on the ``lerna publish`` command to determine which packages have changed since the last merged commit and release, and only publish those. 
+
+Lerna is configured to use independent versioning with conventional commits, as opposed to keeping all package versions in sync. This will allow us to publish packages independently of each other, rather than in lock-step like some JS monorepos do (e.g., Babel, React). The benefit of independent versioning is that consuming applications won't get erroneous upgrades when code unrelated to their use case is updated in the monorepo. For example, say the ``edx/frontend-app-learning`` micro-frontend consumes ``@edx/frontend-enterprise-utils``. Without independent versioning, if we released a new version of ``@edx/frontend-enterprise-catalog-search``, then ``edx/frontend-app-learning`` would have an upgrade available when nothing about the package changed.
+
+The image that follows is a diagram outlining the NPM publishing process. When a PR is merged, Github Actions will run the necessary Lerna commands to run all linting, tests, and builds for each package that changed since the last release, including dependencies and dependents. Each package, independently released, follows semantic versioning to determine the package version, creates and pushes Git tags, and then publishes to NPM.
+
+.. image:: https://user-images.githubusercontent.com/2828721/116780392-82947000-aa4a-11eb-82ce-819604de935f.png
+
+************
+Consequences
+************
+
+* The existing ``frontend-enterprise`` NPM package will be deprecrated in favor of the 3 new NPM packages outlined above. We will need to upgrade all uses of ``frontend-enterprise`` to import from the relevant individual NPM packages instead.
+* Common or shared dependencies will be accessible to all packages in the monorepo, regardless of whether they are explicitly specified in a package's package.json file. Despite this, we have tooling in place through ESlint's ``eslint-plugin-import`` to throw a linting error when attempting to import packages that are not explicitly defined in package.json.
```

Comment:
does this mean that if I introduce a dep into the top level, and a specific package does not need/use it, what will happen? Will it still be in the dependency tree of that package? I was hoping there would be a way to say, only make a dep visible to one of the packages, if they explicitly depend on it, otherwise no.

Reply:
Say we have some new package that doesn't rely on React. That package would not define anything to do with React in its own package.json file, and as such, the NPM distribution won't have any reference or reliance on React; similarly, React wouldn't appear in that package's `node_modules`.

However, because React _is_ installed as a devDep in the root package.json file, it will still be accessible to any package within the monorepo even if it's not explicitly declared in a package's package.json file. This is due to how NPM resolves `node_modules` when importing; it continues to search up the file system tree to find "node_modules" folders in parent directories.

While we can't prevent that import from happening, we will at least throw a linting error (which fails CI) when trying to use a package from the root `node_modules` that isn't explicitly defined in an individual package's package.json file.

See https://github.com/lerna/lerna/blob/main/doc/hoist.md#forgetting-dependencies for more details. This document is specifically about `lerna bootstrap --hoist` which literally takes _all_ shared dependencies across all packages in the monorepo and installs them in the root, but `--hoist` felt too much like magic, especially since we may not want to always "hoist" up all shared dependencies. The only time I think it makes sense to hoist a package is when it will run into conflicts by having 2 different versions in use at the same time (like react, react-dom, etc.). But the same principle applies; we're just sort of being explicit about what to hoist and what not to hoist.